daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -841,11 +841,69 @@ func (devices *DeviceSet) createRegisterDevice(hash string) (*devInfo, error) {
 	return info, nil
 }
 
-func (devices *DeviceSet) createRegisterSnapDevice(hash string, baseInfo *devInfo, size uint64) error {
-	if err := devices.poolHasFreeSpace(); err != nil {
+func (devices *DeviceSet) takeSnapshot(hash string, baseInfo *devInfo, size uint64) error {
+	var (
+		doSuspend      bool
+		devinfo        *devicemapper.Info
+		deferredRemove = true
+		err            error
+	)
+
+	if err = devices.poolHasFreeSpace(); err != nil {
+		return err
+	}
+
+	if devices.deferredRemove {
+		devinfo, err = devicemapper.GetInfoWithDeferred(baseInfo.Name())
+		if err != nil {
+			return err
+		}
+	} else {
+		deferredRemove = false
+		devinfo, err = devicemapper.GetInfo(baseInfo.Name())
+		if err != nil {
+			return err
+		}
+	}
+
+	if devinfo != nil && devinfo.DeferredRemove == 0 {
+		deferredRemove = false
+	}
+
+	if deferredRemove {
+		err = devices.cancelDeferredRemoval(baseInfo)
+		if err != nil {
+			if err != devicemapper.ErrEnxio {
+				return err
+			}
+			// Error is ErrEnxio. Device is probably already gone. Return success.
+			deferredRemove = false
+		}
+	}
+
+	// if deferred removal was cancelled, we can safely assume the device is active.
+	if deferredRemove {
```

Comment:
Given we have already done GetInfo(), I think we can simplify it and not try to infer it from deferred remove. 

doSuspend := devinfo != nil && devinfo.Exists != 0